sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -146,8 +116,99 @@ case class Mode(
   override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): Mode =
     copy(inputAggBufferOffset = newInputAggBufferOffset)
 
-  override def withNewChildrenInternal(newLeft: Expression, newRight: Expression): Expression =
-    copy(child = newLeft, deterministicExpr = newRight)
+  override def sql(isDistinct: Boolean): String = {
+    reverseOpt.map {
+      reverse =>
+        if (reverse) {
+          s"$prettyName() WITHIN GROUP (ORDER BY ${child.sql} DESC)"
+        } else {
+          s"$prettyName() WITHIN GROUP (ORDER BY ${child.sql})"
+        }
+    }.getOrElse(super.sql(isDistinct))
+  }
+
+  override def orderingFilled: Boolean = child != UnresolvedWithinGroup
+
+  assert(orderingFilled || (!orderingFilled && reverseOpt.isEmpty))
+
+  override def withOrderingWithinGroup(orderingWithinGroup: Seq[SortOrder]): AggregateFunction = {
+    child match {
+      case UnresolvedWithinGroup =>
+        if (orderingWithinGroup.length != 1) {
+          throw QueryCompilationErrors.wrongNumOrderingsForInverseDistributionFunctionError(
+            nodeName, 1, orderingWithinGroup.length)
+        }
+        orderingWithinGroup.head match {
+          case SortOrder(child, Ascending, _, _) =>
+            this.copy(child = child, reverseOpt = Some(true))
+          case SortOrder(child, Descending, _, _) =>
+            this.copy(child = child, reverseOpt = Some(false))
+        }
+      // For compatibility with function calls without WITHIN GROUP.
+      case _ =>
+        if (orderingWithinGroup.length != 0) {
```

Comment:
when will we hit this branch? If the child is not `UnresolvedWithinGroup`, then we will fail earlier if `WITHIN GROUP` is specified.